### PR TITLE
Add command scan-proofs

### PR DESCRIPTION
**Pull request by mlsteele**
Adds a dev-only keybase client command `scan-proofs`. This checks whether a client agrees with the server when it comes to checking proofs.

This will be used as:
- A way to check that PVL is working the same as the existing checker.
- A sanity check for the server vs client implementations of proof checking.
- A way to check PVL upgrades (but probably not really, because those will occur when a service breaks our checker).

Here's its usage:
```NAME:
   keybase scan-proofs - Test proof validation on many users.

USAGE:
   keybase scan-proofs [command options]

DESCRIPTION:
   Check that this client and the server agree about a list of proofs.

   Takes as input a file containing a database dump of existing proofs with their server state.
   The input file is a csv of the result of this query:
   https://github.com/keybase/keybase/blob/master/notes/scan-proofs.md

   Using a cache file will remember when successful agreements, speeding up
   subsequent runs of scan-proofs on the same proofs.

   An ignore file can be supplied which contains a list of sig ids to skip checking.
   The ignore file should be a list of sigids, one per line, with optional comments
   starting with "//".

OPTIONS:
   -i, --infile         Specify an input file.
   -w, --indices        Only check in this range of indices. (like 2:8).
   -s, --sigid          Only check this sigid.
   -r, --ratelimit "0"  Check 1 proof every n seconds. Defaults to 0.5
   -c, --cachefile      Specify a file to use as a cache.
   -g, --ignorefile     Specify a file containing a list of sigids to ignore.
```

Basically how it works is:
- read in a file as a list of proofs to check
- for each proof
- run the existing proof checker
- run the pvl proof checker
- compare and report on the 3 results: non-pvl, pvl, server status

And an example (anonymized) run:
```bash
$ keybase-dev --run-mode=prod --home=/tmp/ig --standalone scan-proofs \
  -i ~/Downloads/query_result.csv \
  -c /tmp/checkcache \
  -g /tmp/ignore \
  -r .3 \
  -w 105:110 \

▶ INFO Using cache: /tmp/checkcache (61 entries)
▶ INFO Using ignore file: 1 (/tmp/ignore entries)
▶ INFO Running with ratelimit: 0.3 seconds
▶ INFO i:105 user:anon type:1001 sigid:fakej08hr30h8308h82h083h083h0238hf083hf08h308fh380hf3083hf028fh001
▶ INFO Ok

▶ INFO i:106 user:anon2 type:2 sigid:fakej08hr30h8308h82h083h083h0238hf083hf08h308fh380hf3083hf028fh002
▶ INFO Ok

▶ INFO i:107 user:anon2 type:2 sigid:fakej08hr30h8308h82h083h083h0238hf083hf08h308fh380hf3083hf028fh002
▶ ERROR 49 FAILED: Local verifiers disagree with server: server:3 client:<nil>

▶ INFO i:108 user:anon2 type:2 sigid:fakej08hr30h8308h82h083h083h0238hf083hf08h308fh380hf3083hf028fh002
▶ INFO skipping: server REVOKED
▶ INFO Ok

▶ INFO i:109 user:anon3 type:3 sigid:fakej08hr30h8308h82h083h083h0238hf083hf08h308fh380hf3083hf028fh001
▶ INFO Ok

▶ INFO ---
▶ INFO proofs checked  : 4
▶ INFO oks             : 4
▶ INFO fails           : 1
```

Oh, also, I haven't tried running this without `--standalone`. The logging might all go to the service and not the tty you run it from.

r? @oconnor663 @maxtaco

### CODE REVIEW

File: go/engine/scanproofs.go [hunk 1]
```diff
@@ -0,0 +1,427 @@
...
+		libkb.LogUIKind,
+	}
+}
+
+func (e *ScanProofsEngine) SubConsumers() []libkb.UIConsumer {
+	return []libkb.UIConsumer{}
+}
+
+func (e *ScanProofsEngine) Run(ctx *Context) (err error) {
+	e.G().Log.Debug("+ ScanProofsEngine Run")
```
Comment (maxtaco) [inline go/engine/scanproofs.go]: `defer e.G().Trace("ScanProofEngine#Run", func() error { return err })()`

File: go/engine/scanproofs.go [hunk 2]
```diff
@@ -0,0 +1,427 @@
...
+		skipreason = "server LOOKING"
+	case keybase1.ProofState_SUPERSEDED:
+		skip = true
+		skipreason = "server SUPERSEDED"
+	case keybase1.ProofState_POSTED:
+		badstate = true
+	case keybase1.ProofState_REVOKED:
+		skip = true
+		skipreason = "server REVOKED"
+	case 8: // DELETED
```
Comment (maxtaco) [inline go/engine/scanproofs.go]: can you make a new enum value for this?

**Comment by mlsteele** [comment]
Used trace, added enum.
The enum is also in a separate PR #4123 which I'd land first.

**Comment by maxtaco** [comment]
👍

**Pull request merged**